### PR TITLE
Fix race condition between Breakpoint construction and access from listeners

### PR DESCRIPTION
**Pull request by smarr**
This change fixes a race between the construction of a `BreakpointImpl` object and its access from another listener thread. This can happen easily if another thread related to a debugger UI creates a breakpoint while a program is executing.

Details:
- make BreakpointListener static class, because letting `this` escape to another thread during construction is just evil: I can't see that race when looking at the code
- use factory method for BreakpointImpl construction to register listener after object creation
- this relies on JVM semantics that ensures that all object fields are visible after object construction, in case at least one of the fields in final. Otherwise, we'd need to make the racy fields volatile.

- make assumption fields `@CompilationFinal` otherwise they are not properly optimized

Ping: @chumer @mlvdv

### CODE REVIEW

**Comment by smarr** [comment]
@entlicher this one might be also relevant for the NetBeans debugger.

**Comment by mlvdv** [comment]
looks good.

**Comment by smarr** [comment]
If this fix is acceptable, could please someone accept it and integrate it, @chumer @mlvdv @jtulach?
Thanks

File: truffle/com.oracle.truffle.api.debug/src/com/oracle/truffle/api/debug/BreakpointFactory.java [hunk 1]
```diff
@@ -253,6 +253,16 @@ private void forget(BreakpointImpl breakpoint) {
         breakpoints.remove(breakpoint.getKey());
     }
 
+    BreakpointImpl createBreakpoint(Object key, SourceSectionFilter query, int ignoreCount, boolean isOneShot) {
+        BreakpointImpl breakpoint = new BreakpointImpl(key, query, ignoreCount, isOneShot);
+        // Only register the listener after the breakpoint has been partially
+        // constructed. Otherwise, we can't be sure that the assumptions are visible.
```
Comment (eregon) [inline truffle/com.oracle.truffle.api.debug/src/com/oracle/truffle/api/debug/BreakpointFactory.java]: Why `partially`? The BreakpointImpl will be "fully" constructed, isn't it?

Comment (smarr) [inline truffle/com.oracle.truffle.api.debug/src/com/oracle/truffle/api/debug/BreakpointFactory.java]: will fix the comment

**Comment by smarr** [comment]
Rephrased comment, and rebased on master.

**Comment by entlicher** [comment]
The change looks good to me, thanks.

**Comment by smarr** [comment]
Thanks.

**Pull request merged**